cdap-ui/app/cdap/components/AbstractWidget/SchemaEditor/Context/__tests__/SchemaManager.tests.ts

```diff
@@ -0,0 +1,261 @@
+import {
+  schemaWithMap,
+  schemaWithName,
+  schemaWithModifiedType,
+  simpleSchema,
+  simpleSchema2,
+  largeSchema,
+  largeSchema2,
+} from './schemas';
+import { ISchemaType } from 'components/AbstractWidget/SchemaEditor/SchemaTypes';
+
+const childCountInTree = (tree) => {
+  if (!tree || (tree && !tree.children)) {
+    return 0;
+  }
+  return Object.keys(tree.children).filter((child) => child !== 'order').length;
+};
+
+const getNthFieldIdInFlatSchema = (schemaManagerInstance: ISchemaManager, n) => {
+  const list = schemaManagerInstance.getFlatSchema();
+  const { id, ancestors } = list[n];
+  const fieldId = { id, ancestors };
+  return { fieldId, field: list[n] };
+};
+
+describe('Unit tests for Schema Manager', () => {
+  describe('Basic operations should work', () => {
+    let schema;
+    beforeEach(() => {
+      schema = SchemaManager().getInstance();
+    });
+
+    it('Should work', () => {
+      expect(schema).not.toBeNull();
+      const tree = schema.getSchemaTree();
+      const list = schema.getFlatSchema();
+      expect(childCountInTree(tree)).toBe(1);
+      expect(list.length).toBe(2); // top level schema + first empty record field.
+      expect(list[1].name).toBe('');
+    });
+
+    it('Should change "name" of a record field', () => {
+      const list = schema.getFlatSchema();
+      const { id, ancestors } = list[1];
+      const fieldId = { id, ancestors };
+      schema.onChange(fieldId, {
+        property: 'name',
+        value: 'new_field',
+        type: 'update',
+      });
+      expect(schema.getFlatSchema().length).toBe(2);
+      expect(list[1].name).toBe('new_field');
+      expect(schema.getAvroSchema()).toEqual(schemaWithName);
+    });
+
+    it('Should change the "type" of a record field', () => {
+      const { fieldId } = getNthFieldIdInFlatSchema(schema, 1);
+      schema.onChange(fieldId, {
+        property: 'name',
+        value: 'new_field',
+        type: OperationTypesEnum.UPDATE,
+      });
+      schema.onChange(fieldId, {
+        property: 'type',
+        value: 'boolean',
+        type: OperationTypesEnum.UPDATE,
+      });
+      expect(schema.getFlatSchema()[1].type).toBe('boolean');
+      expect(schema.getAvroSchema()).toEqual(schemaWithModifiedType);
+    });
+
+    it('Should add a new record field', () => {
+      const { fieldId } = getNthFieldIdInFlatSchema(schema, 1);
+      schema.onChange(fieldId, { type: OperationTypesEnum.ADD });
+      expect(schema.getFlatSchema().length).toBe(3);
+    });
+
+    it('Should remove a record field', () => {
+      const { fieldId } = getNthFieldIdInFlatSchema(schema, 1);
+      schema.onChange(fieldId, { type: OperationTypesEnum.ADD });
+      schema.onChange(fieldId, { type: OperationTypesEnum.REMOVE });
+      expect(schema.getFlatSchema().length).toBe(2);
+    });
+
+    it('Should change simple type to complex type', () => {
+      const { fieldId } = getNthFieldIdInFlatSchema(schema, 1);
+      schema.onChange(fieldId, {
```

Comment:
Question, potentially for future work: Would it make sense for `onChange` to accept an array of change structures? Would batching changes lead to performance improvements down the line?

Reply:
Are you considering more of individual field changes or changes across fields? If for individual changes we already sort of isolate the change for performance but it would be a little bit challenging to do across fields.

Say the user enters a new field, say field1, and then changes the type to be int and adds another field. If we were to batch this operation then we temporarily have to store the state somewhere before we update the SchemaManager which then reflects the presentation layer (DOM). I am happy to explore more on the topic.